feat(Man): port to Lua

lewis6991 opened the pull request:
- Mostly a 1-to-1 port from vim; lots of `vim.fn`.
- Works fully AFAIK. Edge case code most definitely has porting bugs.
- Command completion is ~a lot quicker and now usable, but still slow~ now blazingly fast.
- I think bugs in this will be unavoidable so will need to address those in follow up PR's.

lewis6991 committed:
feat(man): port to Lua

justinmk reviewed (approved):
Nice. Keeping it somewhat 1-to-1 definitely will help with comparing previous changes until we feel more confident.

clason commented:
@mhinz

lewis6991 commented:
Wow, no idea why, but the following patch completely fixes completion from being laggy.

```diff
-  local args = fn.split(cmd_line)
+  local args = vim.split(cmd_line, '%s+')
```

smjonas commented:
Nice! Are there any guidelines for which other files can / should be ported to Lua (from a performance or maintainability standpoint)? These "port to Lua" PRs are generally good for new contributors so it would be nice to have a separate issue listing them.

clason commented:
Anything that Vim has ported to vim9script (if that is not merely windowdressing). Current list is here:
https://github.com/neovim/neovim/projects/5#card-75443165

(`matchparen` is soon to be on it, too -- but that would probably profit from a full rewrite to combine it with `matchit`, similarly to how `matchup` does it.)

lewis6991 commented:
Any last comments/reviews before I merge?

clason commented:
LGTM, do you want to squash?

zeertzjq commented on src/nvim/testdir/test_profile.vim:
```diff
@@ -485,11 +485,11 @@ func Test_profdel_func()
+  call assert_equal('Called 1 time',                  lines[2])
+  call assert_equal('FUNCTION  Foo1()',               lines[8])
+  call assert_match('Defined:.*Xprofile_file.vim',    lines[9])
+  call assert_equal('Called 2 times',                 lines[10])
```
Why does this need change?

zeertzjq replied:
Hmm, it seems removing `runtime/plugin/man.vim` changes this. This is strange.

lewis6991 replied:
Yes it is. Very.

zeertzjq commented:
This is a much smaller test change that makes the tests pass:
```vim
diff --git a/src/nvim/testdir/test_profile.vim b/src/nvim/testdir/test_profile.vim
index fdb6f13e2..401ffd396 100644
--- a/src/nvim/testdir/test_profile.vim
+++ b/src/nvim/testdir/test_profile.vim
@@ -40,7 +40,7 @@ func Test_profile_func()
   call writefile(lines, 'Xprofile_func.vim')
   call system(GetVimCommand()
     \ . ' -es --clean'
-    \ . ' -c "so Xprofile_func.vim"'
+    \ . ' --cmd "so Xprofile_func.vim"'
     \ . ' -c "qall!"')
   call assert_equal(0, v:shell_error)
 
@@ -475,7 +475,7 @@ func Test_profdel_func()
     call Foo3()
   [CODE]
   call writefile(lines, 'Xprofile_file.vim')
-  call system(GetVimCommandClean() . ' -es -c "so Xprofile_file.vim" -c q')
+  call system(GetVimCommandClean() . ' -es --cmd "so Xprofile_file.vim" -c q')
   call assert_equal(0, v:shell_error)
 
   let lines = readfile('Xprofile_file.log')
```

lewis6991 commented:
Nice! Do you know why that makes a difference?

zeertzjq commented:
With `-c` the file is sourced after `plugin/` files are loaded, while with `--cmd` the file is sourced before that. Apart from that I have no idea either.

lewis6991 commented:
> LGTM, do you want to squash?

Unless you disagree, I think the commits as they are should be committed. The go brrrr commit is a good bisection point for a perf change without a known reason.

zeertzjq commented:
> feat(mean)

clason commented:
> feat(mean)

That's mean! 😭

lewis6991 commented:
> > The go brrrr commit is a good bisection point for a perf change without a known reason.
> 
> Could do with a more descriptive and less meme-y message, though? (`perf(man): <do foo>`)

Squashed that commit and added a comment instead.

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -1,7 +1,66 @@
+    pcall(function()
+      fn.jobstop(jobid)
+      man_error(string.format('command timed out: %s', table.concat(cmd, ' ')))
+    end)
```
In original code only `E900:` error is suppressed. Other errors are still thrown.

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -1,7 +1,66 @@
+    pcall(function()
+      fn.jobstop(jobid)
+      man_error(string.format('command timed out: %s', table.concat(cmd, ' ')))
+    end)
```
This doesn't look right. What this should do is to do nothing if `jobstop()` throws `E900:` error, and throw a `command timed out` error otherwise.

zeertzjq replied:
Wait, the behavior of `jobstop()` was changed in #11740, so the original code is wrong as well.

lewis6991 replied:
Does it matter though?

zeertzjq replied:
No it doesn't

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +211,529 @@ local function highlight_line(line, linenr)
+    if vim.b.man_default_sects ~= nil then
+      local sects = vim.split(vim.b.man_default_sects, ',', { plain = true })
+      for _, sec in ipairs(sects) do
+        local ok, ret = pcall(get_path, sec, name)
+        if ok then
+          return ret
+        end
```
This should only catch `command error (`, not everything.

zeertzjq replied:
This should only catch `command error`, not everything.

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +211,529 @@ local function highlight_line(line, linenr)
+    end
+  else
+    -- try with specified section
+    local ok, res = pcall(get_path, sect, name)
+
+    if ok and res ~= nil then
+      return res
+    end
```
ditto

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +211,529 @@ local function highlight_line(line, linenr)
+    if vim.b.man_default_sects ~= nil then
+      local sects = vim.split(vim.b.man_default_sects, ',', { plain = true })
+      for _, sec in ipairs(sects) do
+        local ok2, res2 = pcall(get_path, sec, name)
+        if ok2 and res2 ~= nil then
+          return res2
+        end
```
ditto

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +211,529 @@ local function highlight_line(line, linenr)
+  end
+
+  -- if none of the above worked, we will try with no section
+  local ok, res = pcall(get_path, '', name)
+  if ok and res ~= nil then
+    return res
+  end
```
ditto

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +211,529 @@ local function highlight_line(line, linenr)
+  if vim.env.MANSECT then
+    local MANSECT = vim.env.MANSECT
+    vim.env.MANSECT = nil
+    local ok2, res2 = pcall(get_path, '', name)
+    vim.env.MANSECT = MANSECT
+    if ok2 and res2 ~= nil then
+      return res2
+    end
```
ditto

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +218,529 @@ local function highlight_line(line, linenr)
+    }
+  end
+
+  if vim.ocscopetag then
```
```suggestion
  if vim.o.cscopetag then
```

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +218,529 @@ local function highlight_line(line, linenr)
+  if count >= 0 then
+    sect = tostring(count)
+  end
+  verify_exists(sect, name)
```
```suggestion
  local path = verify_exists(sect, name)
  sect, name = extract_sect_and_name_ref(path)
```

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +211,529 @@ local function highlight_line(line, linenr)
+    if not string.find(mods, 'tab') and find_man() then
+      vim.cmd.tag({ target, mods = { silent = true, keepalt = true } })
+    else
+      vim.cmd.stag({ target, mods = { silent = true, keepalt = true } })
```
Add `mods` argument into `mods`

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +211,529 @@ local function highlight_line(line, linenr)
+end
+
+local function init()
+  local ok = pcall(function()
+    -- Check for -l support.
+    get_page(get_path('', 'man'))
+  end)
+
+  if not ok then
+    localfile_arg = false
+  end
```
This should only catch `command error`, not everything.

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +220,529 @@ local function highlight_line(line, linenr)
+    -- It will offer 'priclass.d(1m)' even though section is specified as 1.
+    local tmp = vim.split(arg_lead, '(', { plain = true })
+    name = tmp[1]
+    sect = tmp[2]:lower()
```
```suggestion
    sect = (tmp[2] or ''):lower()
```

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +220,529 @@ local function highlight_line(line, linenr)
+-- see extract_sect_and_name_ref on why tolower(sect)
+function M.man_complete(arg_lead, cmd_line, _)
+  -- TODO(lewis6991): for some reason using vim.fn.split here makes completion very laggy
+  local args = vim.split(cmd_line, '%s+')
```
To equivalent to `vim.fn.split()`, `trimempty = true`. It make a difference here. Not sure about other places, but I guess it is better to add `trimempty = true` to other places as well.
```suggestion
  local args = vim.split(cmd_line, '%s+', {trimempty = true})
```

zeertzjq replied:
Currently completion results contain empty strings. When I tried adding `trimempty = true` to every `vim.split()` empty strings no longer appear.

lewis6991 replied:
Ahhh! This explains why the old implementation was really laggy. If there was no arg_lead then `args == {'Man'}` which caused `complete('', '', '')` to be called which is really slow. Without `trimempty` then the `#args==1` branch is unreachable.

lewis6991 commented on runtime/lua/man.lua:
```diff
@@ -152,21 +220,529 @@ local function highlight_line(line, linenr)
+    -- Fallback to a single path, with the page we're trying to find.
+    ok, ret = pcall(verify_exists, sect, name)
+
+    return { ok and ret or nil }
```
We don't want to return a list of a list here.

If `ok` we return `{ ret }`, if `not ok` we return `{}`.

lewis6991 committed:
fixup!: apply suggestions from code review

Co-authored-by: zeertzjq <zeertzjq@outlook.com>

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +220,539 @@ local function highlight_line(line, linenr)
+    return nil
+  end
+
+  local results = vim.split(lines, '\n')
```
```suggestion
  local results = vim.split(lines, '\n', { trimempty = true })
```

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +220,539 @@ local function highlight_line(line, linenr)
+local function get_paths(sect, name, do_fallback)
+  -- callers must try-catch this, as some `man` implementations don't support `s:find_arg`
+  local ok, ret = pcall(function()
+    local mandirs = table.concat(vim.split(man_system({ 'man', find_arg }), '[:\n]'), ',')
```
```suggestion
    local mandirs = table.concat(vim.split(man_system({ 'man', find_arg }), '[:\n]', { trimempty = true }), ',')
```

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +220,539 @@ local function highlight_line(line, linenr)
+
+-- see extract_sect_and_name_ref on why tolower(sect)
+function M.man_complete(arg_lead, cmd_line, _)
+  local args = vim.split(cmd_line, '%s+', {trimempty = true})
```
```suggestion
  local args = vim.split(cmd_line, '%s+', { trimempty = true })
```

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +220,539 @@ local function highlight_line(line, linenr)
+
+  local ok, ret = pcall(function()
+    if not smods.tab and find_man() then
+      vim.cmd.tag({ target, mods = { silent = true, keepalt = true } })
+    else
+      smods.silent = true
+      smods.keepalt = true
+      vim.cmd.stag({ target, mods = smods })
```
Hmm, I just found a bug in `nvim_create_user_command`.

zeertzjq commented on runtime/lua/man.lua:
```diff
@@ -152,21 +220,539 @@ local function highlight_line(line, linenr)
+
+  local ok, ret = pcall(function()
+    if not smods.tab and find_man() then
+      vim.cmd.tag({ target, mods = { silent = true, keepalt = true } })
+    else
+      smods.silent = true
+      smods.keepalt = true
+      vim.cmd.stag({ target, mods = smods })
```
Hmm, I just found a bug in `nvim_create_user_command` and `nvim_parse_cmd`.

zeertzjq replied:
#20039

aktau commented:
Just recompiled from HEAD, and I get the following:

```
$ echo $MANPAGER
nvim +Man!
$ man open
Error detected while processing command line..function man#init_pager:                                                                                                                                                                      
line    6:                                                                                                                                                                                                                                  
E5108: Error executing lua [string ":lua"]:1: attempt to call field 'highlight_man_page' (a nil value)                                                                                                                                      
stack traceback:                                                                                                                                                                                                                            
        [string ":lua"]:1: in main chunk 
```

lewis6991 commented:
I think you need to delete `man.vim` from your installed runtime files. Better to just delete all nvim runtime and install again.

clason commented:
PSA: `make install` does not _remove_ installed runtime files. You need to do that manually.

aktau commented:
Thanks for the PSA. I should've known that, but didn't. The fix worked.
